include/openmc/source.h

```diff
@@ -19,32 +19,47 @@ namespace openmc {
 // Global variables
 //==============================================================================
 
-class SourceDistribution;
+class Source;
 
 namespace model {
 
-extern std::vector<SourceDistribution> external_sources;
+extern std::vector<std::unique_ptr<Source>> external_sources;
 
 } // namespace model
 
 //==============================================================================
-//! External source distribution
+//! Abstract source interface
 //==============================================================================
 
-class SourceDistribution {
+class Source {
+public:
+  virtual ~Source() = default;
+
+  // Methods that must be implemented
+  virtual Particle::Bank sample(uint64_t* seed) = 0;
```

Comment:
```suggestion
  virtual Particle::Bank sample(uint64_t* seed) const = 0;
```

Perhaps this pure virtual method, and the implementations of it, should be declared `const`? I don't think we want to be (or indeed potentially allowing users with a custom source to be) modifying the instance when sampling. I think that was an omission on my part when originally putting the first pass at this abstract base class together.